tests/end_to_end/features/steps/autovariants.py

```diff
@@ -0,0 +1,132 @@
+
+    context.variant = ff.get_run()
+
+
+@when("I register a transformation with auto variant")
+def step_impl(context):
+    @context.spark.df_transformation(
+        inputs=[context.transactions],
+    )
+    def some_transformation(df):
+        """Unedited transactions"""
+        return df
+
+    context.client.apply(asynchronous=False, verbose=True)
+    context.transformation = some_transformation
+
+
+@then("I should be able to reuse the same variant for the same transformation")
+def step_impl(context):
+    @context.spark.df_transformation(
+        inputs=[context.transactions],
+    )
+    def some_transformation(df):
+        """Unedited transactions"""
+        return df
+
+    context.client.apply(asynchronous=False, verbose=True)
+    context.new_transformation = some_transformation
+
+    assert (
+        context.new_transformation.name_variant()
+        == context.transformation.name_variant()
+    )
+
+
+@then("I should be able to register a new auto variant transformation")
+def step_impl(context):
+    @context.spark.df_transformation(
+        inputs=[context.transactions],
+    )
+    def some_transformation(df):
+        """Unedited transactions"""
+        return df
+
+    context.client.apply(asynchronous=False, verbose=True)
+    context.new_transformation = some_transformation
+
+    assert (
+        context.new_transformation.name_variant()
+        != context.transformation.name_variant()
+    )
+
+
+@then("I can get the transformation as df")
+def step_impl(context):
+    df = context.client.dataframe(context.new_transformation)
+    assert len(df) > 0
+
+
+@when("I register a transformation with user-provided variant")
+def step_impl(context):
+    new_random_variant = f"v{str(random.randint(10000, 99999))}"
+
+    @context.spark.df_transformation(
+        variant=new_random_variant,
+        inputs=[context.transactions],
+    )
+    def some_transformation(df):
+        """Unedited transactions"""
+        return df
+
+    context.client.apply(asynchronous=False, verbose=True)
+    context.transformation = some_transformation
+
+
+@then("I should be able to register a modified transformation with new auto variant")
+def step_impl(context):
+    @context.spark.df_transformation(
+        inputs=[context.transactions],
+    )
+    def some_transformation(df):
+        """modified transactions"""
+        from pyspark.sql.functions import lit
+
+        df = df.withColumn("new_column", lit("hi.world"))
+        return df
+
+    context.client.apply(asynchronous=False, verbose=True)
+    context.new_transformation = some_transformation
+
+    assert (
+        context.new_transformation.name_variant()[0]
+        == context.transformation.name_variant()[0]
+    )
+    assert (
+        context.new_transformation.name_variant()[1]
+        != context.transformation.name_variant()[1]
+    )
+
+
+@then("I should be able to register a source with user-defined variant")
```

Comment:
what do you think about "register file (with variant {variant}" or something more reusable